Spigot-Server-Patches/0696-Skip-unnecessary-hopper-pulls-and-pushes.patch

```diff
@@ -0,0 +1,279 @@
+                 this.world.updateAdjacentComparators(this.position, this.c.getBlock());
+             }
++        // Paper start - keep track of inventory updates
++        } else {
++            if (this instanceof IInventory) {
++                Chunk.lastUnknownInventoryUpdate = MinecraftServer.currentTick;
++            }
++        // Paper end
+         }
+ 
+     }
+diff --git a/src/main/java/net/minecraft/world/level/block/entity/TileEntityHopper.java b/src/main/java/net/minecraft/world/level/block/entity/TileEntityHopper.java
+index 537dc52e5ff3325555ee6049bc7f277952983b76..63a419c676617c445f233ed03c753df3bff42f36 100644
+--- a/src/main/java/net/minecraft/world/level/block/entity/TileEntityHopper.java
++++ b/src/main/java/net/minecraft/world/level/block/entity/TileEntityHopper.java
+@@ -12,6 +12,7 @@ import net.minecraft.core.NonNullList;
+ import net.minecraft.nbt.NBTTagCompound;
+ import net.minecraft.network.chat.ChatMessage;
+ import net.minecraft.network.chat.IChatBaseComponent;
++import net.minecraft.server.MinecraftServer;
+ import net.minecraft.world.ContainerUtil;
+ import net.minecraft.world.IInventory;
+ import net.minecraft.world.IInventoryHolder;
+@@ -29,6 +30,7 @@ import net.minecraft.world.level.block.Block;
+ import net.minecraft.world.level.block.BlockChest;
+ import net.minecraft.world.level.block.BlockHopper;
+ import net.minecraft.world.level.block.state.IBlockData;
++import net.minecraft.world.level.chunk.Chunk;
+ import net.minecraft.world.phys.AxisAlignedBB;
+ import net.minecraft.world.phys.shapes.OperatorBoolean;
+ import net.minecraft.world.phys.shapes.VoxelShapes;
+@@ -159,13 +161,15 @@ public class TileEntityHopper extends TileEntityLootable implements IHopper, ITi
+             if (!this.m() && (Boolean) this.getBlock().get(BlockHopper.ENABLED)) {
+                 boolean flag = false;
+ 
+-                if (!this.isEmpty()) {
++                if (!shouldSkipPush() && !this.isEmpty()) { // Paper - don't attempt a futile push
+                     flag = this.k();
+                 }
++                lastAttemptedPush = MinecraftServer.currentTick; // Paper
+ 
+-                if (!this.j()) {
++                if (!shouldSkipPull() && !this.j()) { // Paper - don't attempt a futile pull
+                     flag |= (Boolean) supplier.get();
+                 }
++                lastAttemptedPull = MinecraftServer.currentTick; // Paper
+ 
+                 if (flag) {
+                     this.setCooldown(world.spigotConfig.hopperTransfer); // Spigot
+@@ -221,6 +225,7 @@ public class TileEntityHopper extends TileEntityLootable implements IHopper, ITi
+                     itemstack = callPushMoveEvent(iinventory, itemstack);
+                     if (itemstack == null) { // cancelled
+                         origItemStack.setCount(origCount);
++                        lastSuccessfulPush = MinecraftServer.currentTick; // Count as success because plugins are unpredictable
```

Comment:
// Paper

Reply:
It's not clear from the patch, but all the lines missing the // Paper are between // Paper start and // Paper end statements from a different patch